pkg/controller/pxc/users.go

```diff
@@ -317,160 +510,166 @@ func (r *ReconcilePerconaXtraDBCluster) grantSystemUserPrivilege(cr *api.Percona
-		addr = cr.Name + "-pxc-unready." + cr.Namespace + ":33062"
-	}
 	um, err := users.NewManager(addr, pxcUser, pxcPass, cr.Spec.PXC.ReadinessProbes.TimeoutSeconds)
 	if err != nil {
-		return res, errors.Wrap(err, "new users manager")
+		return errors.Wrap(err, "new users manager")
 	}
 	defer um.Close()
 
-	err = um.UpdateUsersPass(sysUsers)
+	err = um.CreateReplicationUser(string(pass))
+	if err != nil {
+		return errors.Wrap(err, "create replication user")
+	}
+
+	sysUsersSecretObj.Data["replication"] = pass
+	internalSysSecretObj.Data["replication"] = pass
+
+	err = r.client.Update(context.TODO(), sysUsersSecretObj)
 	if err != nil {
-		return res, errors.Wrap(err, "update sys users pass")
+		return errors.Wrap(err, "update sys users secret")
+	}
+	err = r.client.Update(context.TODO(), internalSysSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "update internal users secret")
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) handleProxyadminUser(cr *api.PerconaXtraDBCluster, secrets, internalSecrets *corev1.Secret, upRes *userUpdateActions) error {
+	if cr.Spec.ProxySQL != nil && cr.Spec.ProxySQL.Enabled {
+		return errors.New("ProxySQL not enabled")
 	}
+
+	user := &users.SysUser{
+		Name: users.UserProxyAdmin,
+		Pass: string(secrets.Data[users.UserProxyAdmin]),
+	}
+
+	// update proxy users
 	if cr.Spec.ProxySQL != nil && cr.Spec.ProxySQL.Enabled {
-		err = r.updateProxyUsers(proxyUsers, internalSysSecretObj, cr)
+		err := r.updateProxyUser(cr, internalSecrets, user)
 		if err != nil {
-			return res, errors.Wrap(err, "update Proxy users pass")
+			return errors.Wrap(err, "update Proxy users")
 		}
 	}
-	if todo&syncProxyUsers != 0 && !res.restartProxy {
-		err = r.syncPXCUsersWithProxySQL(cr)
-		if err != nil {
-			return res, errors.Wrap(err, "sync users")
-		}
+
+	// syncProxyUser
+	err := r.syncPXCUsersWithProxySQL(cr)
+	if err != nil {
+		return errors.Wrap(err, "sync proxy users")
 	}
 
-	return res, nil
+	//update internalSecrets
+	internalSecrets.Data[users.UserRoot] = secrets.Data[users.UserRoot]
+	err = r.client.Update(context.TODO(), internalSecrets)
+	if err != nil {
+		return errors.Wrap(err, "update internal users secrets proxyadmin user password")
+	}
+
+	upRes.restartProxy = true
+	if cr.Spec.PMM != nil && cr.Spec.PMM.Enabled {
+		upRes.restartPXC = true
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) handlePMMUser(cr *api.PerconaXtraDBCluster, secrets, internalSecrets *corev1.Secret, upRes *userUpdateActions) error {
+	if cr.Spec.PMM != nil && cr.Spec.PMM.Enabled {
+		return errors.New("PMM not enabled")
+	}
+
+	name := users.UserPMMServerKey
+	if !cr.Spec.PMM.UseAPI(secrets) {
+		name = users.UserPMMServer
+	}
+	user := &users.SysUser{
+		Name: name,
+		Pass: string(secrets.Data[name]),
+	}
+
+	// update pass
+	err := r.updateUserPass(cr, secrets, internalSecrets, user)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("update %s users pass", name))
+	}
```

Comment:
do we have PMM server user in the DB?

Reply:
I just checked I don't see them. So we should not do this operation (in original impl it is not distinguished).